kraken-java-sdk/kraken-java-sdk-controller/src/main/java/com/consoleconnect/kraken/operator/controller/api/v3/TemplateUpgradeV3Controller.java

```diff
@@ -91,4 +105,79 @@ public HttpResponse<List<ComponentExpandDTO>> listApiUseCases(
+  public HttpResponse<Paging<TemplateUpgradeReleaseVO>> listTemplateChangeLog(
+      @PathVariable("productId") String productId,
+      @RequestParam(value = "templateUpgradeId", required = false) String templateUpgradeId,
+      @RequestParam(value = "orderBy", required = false, defaultValue = "createdAt") String orderBy,
+      @RequestParam(value = "direction", required = false, defaultValue = "DESC")
+          Sort.Direction direction,
+      @RequestParam(value = "page", required = false, defaultValue = PagingHelper.DEFAULT_PAGE_STR)
+          int page,
+      @RequestParam(value = "size", required = false, defaultValue = PagingHelper.DEFAULT_SIZE_STR)
+          int size) {
+    List<Tuple2> tuple3s = List.of();
+    Paging<UnifiedAssetDto> assetDtoPaging =
+        unifiedAssetService.findBySpecification(
+            Tuple2.ofList(
+                AssetsConstants.FIELD_KIND,
+                AssetKindEnum.PRODUCT_TEMPLATE_UPGRADE.getKind(),
+                AssetsConstants.FIELD_ID,
+                templateUpgradeId),
+            tuple3s,
+            null,
+            getSearchPageRequest(page, size, direction, orderBy),
+            null);
+    List<TemplateUpgradeReleaseVO> list =
+        assetDtoPaging.getData().stream()
+            .map(templateUpgradeService::generateTemplateUpgradeReleaseVO)
+            .toList();
+    // the latest can upgrade
+    list.stream().findFirst().ifPresent(vo -> this.calculateStatus(vo, true));
+    list.stream().skip(1).forEach(vo -> this.calculateStatus(vo, false));
+    return HttpResponse.ok(
+        PagingHelper.toPageNoSubList(
+            list, assetDtoPaging.getPage(), assetDtoPaging.getSize(), assetDtoPaging.getTotal()));
+  }
+
+  protected void calculateStatus(TemplateUpgradeReleaseVO vo, boolean first) {
+    boolean deployedProduction =
+        vo.getDeployments().stream()
+            .anyMatch(
+                deploy -> deploy.getEnvName().equalsIgnoreCase(EnvNameEnum.PRODUCTION.name()));
+    if (deployedProduction) {
+      vo.setStatus(UPGRADED);
+      return;
+    }
+    if (first) {
+      if (CollectionUtils.isEmpty(vo.getDeployments())) {
+        vo.setStatus(NOT_UPGRADED);
+        return;
+      }
+      vo.setStatus(UPGRADING);
+    } else {
+      vo.setStatus(DEPRECATED);
+    }
+  }
+
+  @Operation(summary = "stage environment upgrade check")
+  @GetMapping("/stage-upgrade-check")
+  public HttpResponse<List<String>> checkStageUpgradeCondition(
+      @PathVariable("productId") String productId,
+      @RequestParam(value = "templateUpgradeId", required = false) String templateUpgradeId,
+      @RequestParam String stageEnvId) {
+    return HttpResponse.ok(templateUpgradeService.stageCheck(templateUpgradeId, stageEnvId));
+  }
+
+  @Operation(summary = "product environment upgrade check")
+  @GetMapping("/production-upgrade-check")
+  public HttpResponse<List<String>> checkProductionUpgradeCondition(
+      @PathVariable("productId") String productId,
```

Comment:
the endpoint already indicates this is for `production`, but `productEnvId` is still required.

Reply:
rename parameter productEnvId  to envId